// @sap-cloud-sdk/connectivity/dist/scp-cf/client-credentials-token-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientCredentialsTokenCache = void 0;
exports.getCacheKey = getCacheKey;
exports.getIasCacheKey = getIasCacheKey;
const util_1 = require("@sap-cloud-sdk/util");
const cache_1 = require("./cache");
const logger = (0, util_1.createLogger)({
    package: 'connectivity',
    messageContext: 'client-credentials-token-cache'
});
const ClientCredentialsTokenCache = (cache) => ({
    getToken: (tenantId, clientId) => cache.get(getCacheKey(tenantId, clientId)),
    cacheToken: (tenantId, clientId, token) => {
        cache.set(getCacheKey(tenantId, clientId), {
            entry: token,
            expires: token.expires_in
                ? Date.now() + token.expires_in * 1000
                : undefined
        });
    },
    getTokenIas: (data) => cache.get(getIasCacheKey(data)),
    cacheIasToken: (data, tokenResponse) => {
        cache.set(getIasCacheKey(data), {
            entry: tokenResponse,
            expires: tokenResponse.expires_in
                ? Date.now() + tokenResponse.expires_in * 1000
                : undefined
        });
    },
    clear: () => {
        cache.clear();
    },
    getCacheInstance: () => cache
});
/**
 * Normalizes the IAS resource parameter to a consistent string format for cache key.
 * @param resource - The resource parameter from iasOptions.
 * @returns Normalized resource string or empty string if not provided.
 * @internal
 */
function normalizeResource(resource) {
    if (!resource) {
        return [];
    }
    if ('name' in resource) {
        return [`name=${resource.name}`];
    }
    const normalized = [
        `provider-clientId=${resource.providerClientId}`
    ];
    if (resource.providerTenantId) {
        normalized.push(`provider-tenantId=${resource.providerTenantId}`);
    }
    return normalized;
}
/** *
 * @internal
 * @param tenantId - The ID of the tenant to cache the token for.
 * @param clientId - ClientId to fetch the token.
 * @returns The cache key.
 */
function getCacheKey(tenantId, clientId) {
    if (!tenantId) {
        logger.warn('Cannot create cache key for client credentials token cache. The given tenant ID is undefined.');
        return;
    }
    if (!clientId) {
        logger.warn('Cannot create cache key for client credentials token cache. The given client ID is undefined.');
        return;
    }
    return [tenantId, clientId].join(':');
}
/** *
 * @internal
 * @param data.iasInstance - The IAS instance (tenant) hostname the token is fetched from.
 * @param data.appTid - The BTP instance (tenant) id (App-Tid)
 * @param data.clientId - ClientId to fetch the token.
 * @param data.resource - The App-To-App resource the token is scoped for.
 * @returns The cache key.
 */
function getIasCacheKey(data) {
    const { iasInstance, appTid, clientId, resource } = data;
    if (!iasInstance) {
        logger.warn('Cannot create cache key for client credentials token cache. The given IAS instance hostname is undefined.');
        return;
    }
    if (!clientId) {
        logger.warn('Cannot create cache key for client credentials token cache. The given client ID is undefined.');
        return;
    }
    const output = [
        iasInstance,
        appTid || '',
        clientId,
        ...normalizeResource(resource)
    ];
    return output.join(':');
}
/**
 * @internal
 */
exports.clientCredentialsTokenCache = ClientCredentialsTokenCache(new cache_1.Cache(5 * 60 * 1000 /* 5 minutes in ms */));
