// @sap-cloud-sdk/http-client/dist/compress-request-middleware.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.compress = compress;
const zlib = __importStar(require("node:zlib"));
const node_util_1 = require("node:util");
const util_1 = require("@sap-cloud-sdk/util");
const logger = (0, util_1.createLogger)({
    package: 'http-client',
    messageContext: 'compress-request-middleware'
});
const brotliCompress = (0, node_util_1.promisify)(zlib.brotliCompress);
const gzipCompress = (0, node_util_1.promisify)(zlib.gzip);
const deflateCompress = (0, node_util_1.promisify)(zlib.deflate);
const zstdCompress = zlib.zstdCompress
    ? (0, node_util_1.promisify)(zlib.zstdCompress)
    : undefined;
const compressors = {
    brotli: brotliCompress,
    gzip: gzipCompress,
    deflate: deflateCompress,
    zstd: zstdCompress
};
function getSupportedAlgorithms() {
    return Object.entries(compressors)
        .filter(([, fn]) => fn !== undefined)
        .map(([name]) => `'${name}'`)
        .join(', ');
}
/**
 * Determines whether the payload needs compression based on the provided options.
 * @param payload - The HTTP request payload.
 * @param options - Configuration options for request compression.
 * @returns Returns if the payload should be compressed.
 */
function checkIfNeedsCompression(payload, options) {
    const mode = options?.mode ?? 'auto';
    if (mode === 'header-only' || mode === 'never') {
        return false;
    }
    if (mode === 'always') {
        return true;
    }
    // Ensure we are in 'auto' mode if we reach this point
    mode;
    let payloadSize;
    try {
        payloadSize = Buffer.byteLength(payload);
    }
    catch (e) {
        logger.error(new util_1.ErrorWithCause("Could not determine payload size for 'auto' compression decision. Payload will not be compressed.", e));
        // Skip compression if payload size cannot be determined.
        return false;
    }
    const minSize = options?.autoCompressMinSize ?? 1024;
    const shouldCompress = payloadSize >= minSize;
    const comparison = shouldCompress ? '>=' : '<';
    const action = shouldCompress ? 'Compressing' : 'Skipping compression';
    logger.debug(`Auto compression: payload size ${payloadSize} bytes ${comparison} threshold ${minSize} bytes. ${action}.`);
    return shouldCompress;
}
function getContentEncodingValue(algorithm) {
    switch (algorithm) {
        case 'brotli':
            return 'br';
        case undefined:
        case 'gzip':
            return 'gzip';
        default:
            return algorithm;
    }
}
/**
 * Middleware to compress HTTP request payloads.
 * @param options - Configuration options for request compression.
 * @remarks
 * **Middleware Ordering**: Place compression middleware early in your middleware array,
 * typically after logging/csrf but before resilience middleware (retry, timeout, circuit breaker).
 * This ensures the payload is compressed once and reused across retry attempts.
 * @returns An HTTP middleware that compresses request payloads based on the provided options.
 */
function compress(options) {
    return (middlewareOptions) => async (requestConfig) => {
        const algorithm = options?.algorithm ?? 'gzip';
        const needsCompression = checkIfNeedsCompression(requestConfig.data, options);
        const mode = options?.mode ?? 'auto';
        if (!needsCompression && mode !== 'header-only') {
            return middlewareOptions.fn(requestConfig);
        }
        // If the payload already has a Content-Encoding header, we need to preserve it and append our encoding to it.
        const currentContentEncoding = (0, util_1.pickValueIgnoreCase)(requestConfig.headers, 'content-encoding');
        const algorithmContentEncoding = getContentEncodingValue(algorithm);
        const targetContentEncoding = currentContentEncoding
            ? `${currentContentEncoding}, ${algorithmContentEncoding}`
            : algorithmContentEncoding;
        requestConfig.headers = (0, util_1.mergeIgnoreCase)(requestConfig.headers, {
            'content-encoding': targetContentEncoding
        });
        if (mode === 'header-only') {
            return middlewareOptions.fn(requestConfig);
        }
        const compressor = compressors[algorithm];
        if (!compressor) {
            if (algorithm === 'zstd') {
                throw new Error(`'zstd' compression is not supported in Node.js versions older than v22.15.0. Supported algorithms for this version are: ${getSupportedAlgorithms()}.`);
            }
            throw new Error(`Unsupported compression algorithm '${algorithm}'. Supported algorithms are: ${getSupportedAlgorithms()}.`);
        }
        // TODO: (future) Consider streaming compression for large payloads
        const compressed = await compressor(requestConfig.data, options?.compressOptions).catch((err) => {
            throw new util_1.ErrorWithCause(`Failed to compress request payload using '${algorithm}'.`, err);
        });
        requestConfig.data = compressed;
        return middlewareOptions.fn(requestConfig);
    };
}
